src/core/model/Monitor.h

```diff
@@ -0,0 +1,44 @@
+#include <cassert>
+#include <mutex>
+
+template<class T>
+class Monitor
+{
+public:
+    template<typename ...Args>
+    Monitor(Args&&... args);
+
+    struct LockedMonitor
+    {
+        LockedMonitor(Monitor* monitor);
+        T* operator->();
+        void ReplaceModel(T model);
+        private:
+            Monitor* mon;
+            std::unique_lock<std::mutex> lock;
+    };
+
+    struct TryLockedMonitor
+    {
+        TryLockedMonitor(Monitor* monitor);
+        T* operator->();
+        void ReplaceModel(T model);
+        bool lockAcquired;
+        private:
+            Monitor* mon = nullptr;
+            std::unique_lock<std::mutex> lock;
+    };
+
+
+    LockedMonitor operator->();
+
```

Comment:
Add 
```
SharedMonitor lock_shared();
std::optional<SharedMonitor> try_lock_shared();
```

Reply:
Do we really need this differentiation? This will add a lot of complexity.

When we add something like that the caller is again responsible for only doing read operations while holding the shared monitor. We have no way of forcing a uniquely locked monitor for some function calls to Document.

Reply:
When the `SharedMonitor` only returns a `Document const*` the caller can't access non const functions.